Pass over the outlining section (#620)
browserengineering/book: book/accessibility.md

                self.focus_element(elt)
                self.activate_element(elt)
                self.set_needs_render()
                return
            elt = elt.parent
```

Note that hitting `Enter` when focused on a text entry clears the text
entry; in most browsers, it submits the containing form instead. That
quirk is because [our browser doesn't implement][clear-input] the
`Backspace` key.

[clear-input]: forms.html#interacting-with-widgets

Finally, note that sometimes activating an element submits a form or
navigates to a new page, which means the element we were focused on no
longer exists. We need to make sure to clear focus in this case:

``` {.python}
class Tab:
    def load(self, url, body=None):
        self.focus = None
        # ...
```

We now have the ability to focus on links, buttons, and text entries.
But as with any browser feature, it's worth asking whether web page
authors should be able to customize it. With keyboard navigation, the
author might want certain links not to be focusable (like "permalinks"
to a section heading, which would just be noise to most users), or
might want to change the order in which the user tabs through
focusable items.

> Browsers support the `tabindex` HTML attribute to make this possible.

Author: I rephrased the next sentence on focusability to stand out a bit more, and I've implemented the -1 no focus thing.

``` {.python}
def get_tabindex(node):
    return int(node.attributes.get("tabindex", "9999999"))
```

The default value, "9999999", makes sure that elements without a
`tabindex` attribute sort after ones with the attribute. Now we can
sort by `get_tabindex` in `advance_tab`:

``` {.python}
class Tab:
    def advance_tab(self):
        focusable_nodes = [node
            for node in tree_to_list(self.nodes, [])
            if isinstance(node, Element) and is_focusable(node)
            and get_tabindex(node) >= 0]
        focusable_nodes.sort(key=get_tabindex)
        # ...
```

Since Python's sort is "stable", two elements with the same `tabindex`
won't change their relative position in `focusable_nodes`.

Additionally, elements with `tabindex` are automatically focusable,
even if they aren't a link or a button or a text entry. That's useful,
because that element might listen to the `click` event. To support
this let's first extend `is_focusable` to consider `tabindex`:

``` {.python}
def is_focusable(node):
    return node.tag in ["input", "button", "a"] \
        or "tabindex" in node.attributes
```

Next, we need to make sure to send a `click` event when an element is

> activated:

Reviewer: Add a sentence about the JS behavior, since just like click  enter can be preventDefaulted

Indicating focus
================

Thanks to our keyboard shortcuts, users can now reach any link,
button, or text entry from the keyboard. But if you try to use this to
navigate a website, it's a little hard to know which element is
focused when. A visual indication---similar to the cursor we use on
text inputs---would help sighted users know if they've reached the
element they want or if they need to keep hitting `Tab`. In most
browsers, this visual indiciation is done with a *focus ring* that
outlines the focused element.

To implement focus rings, we're going to have to generalize how we
implemented text cursors. Recall that, right now, text cursors are
added by drawing a verticle line in the `Tab`'s `render` method. We
could extend that code to draw a cursor or an outline, but before we
make that method more complicated let's move it into the `InputLayout`
so we have easier access to size and position information.[^effects]
To do that, we'll need each `InputLayout` to know whether or not it is
currently focused:

[^effects]: As a side effect, this change will also mean text cursors
    are now affected by visual effects, including blends, opacity, and

>     translations. Translations in particular are important.

Author: Yes, as of Chapter 13. More generally, however, I think it tells me that _way back_ in Chapter 7, we should have each `InputLayout` draw its text cursor in `paint` by consulting an `is_focused` field (like we now do in this chapter). I think at the time we checked Chrome and it does it the way we do it in the book, but that looks like a mistake now.

        paint_outline(self.node, cmds, rect)
        cmds = paint_visual_effects(self.node, cmds, rect)
        display_list.extend(cmds)
```

Note that this comes after painting the rest of the text entry's
content but before the visual effects.

Unfortunately, handling links is a little more complicated. That's
because one `<a>` element corresponds to multiple `TextLayout`
objects, so there's not just one layout object where we can stick the
code. Moreover, those `TextLayout`s could be split across several
lines, so we might want to draw more than one focus ring. To work
around this, let's draw the focus ring in `LineLayout`. Each
`LineLayout` finds all of its child `TextLayout`s that are focused,
and draws a rectangle around them all:

``` {.python expected=False}
class LineLayout:
    def paint(self, display_list):
        # ...
        outline_rect = skia.Rect.MakeEmpty()
        parent = None
        for child in self.children:
            parent = child.node.parent
            if has_outline(parent):
                outline_node = parent
                outline_rect.join(child.rect())
        if parent:
            paint_outline(parent, display_list, outline_rect)
```

> You can also add a `paint_outline` call to `BlockLayout`,

Reviewer: ```suggestion
You should also add a `paint_outline` call to `BlockLayout`,
```

:::

Next up is customizing the focus rectangle via the [`outline`][outline] CSS
property. As usual, we'll implement only a subset of the full definition; in
particular, syntax that looks like this:

[outline]: https://developer.mozilla.org/en-US/docs/Web/CSS/outline

    outline: 3px solid red;

> Which means "make the outline red 3px thick". First parse it:

Reviewer: ```suggestion
Which means "make the outline red and 3 pixels thick". First parse it:
```

``` {.python}
def parse_outline(outline_str):
    if not outline_str:
        return None
    values = outline_str.split(" ")
    if len(values) != 3:
        return None
    if values[1] != "solid":
        return None
    return (int(values[0][:-2]), values[2])
```

Now to use it. The outline will be present if the element is focused or has the

> `outline` CSS property generally. While we could finish implementing that via

Reviewer: ```suggestion
`outline` CSS property generally. However: while we could finish implementing that via
```
